packages/puppeteer-core/src/common/Page.ts

```diff
@@ -253,68 +170,225 @@ export class CdpPage extends Page {
     this.#screenshotTaskQueue = screenshotTaskQueue;
     this.#viewport = null;
 
-    this.#setupEventListeners();
-
-    this.#tabSession?.on(CDPSessionEvent.Swapped, async newSession => {
-      this.#client = newSession;
-      assert(
-        this.#client instanceof CdpCDPSession,
-        'CDPSession is not instance of CDPSessionImpl'
-      );
-      this.#target = this.#client._target();
-      assert(this.#target, 'Missing target on swap');
-      this.#keyboard.updateClient(newSession);
-      this.#mouse.updateClient(newSession);
-      this.#touchscreen.updateClient(newSession);
-      this.#accessibility.updateClient(newSession);
-      this.#emulationManager.updateClient(newSession);
-      this.#tracing.updateClient(newSession);
-      this.#coverage.updateClient(newSession);
-      await this.#frameManager.swapFrameTree(newSession);
-      this.#setupEventListeners();
-    });
-    this.#tabSession?.on(CDPSessionEvent.Ready, session => {
-      assert(session instanceof CdpCDPSession);
-      if (session._target()._subtype() !== 'prerender') {
-        return;
-      }
-      this.#frameManager.registerSpeculativeSession(session).catch(debugError);
-      this.#emulationManager
-        .registerSpeculativeSession(session)
-        .catch(debugError);
+    this.once(PageEvent.Close, () => {
```

Comment:
You can do this before emitting the event rather then waiting for it here.

Reply:
Is there a difference? I plan to refactor this in any case. Not here.

Reply:
It's related to order, but also setting up one extra listener that is not needed.